homeassistant/components/jellyfin/browse_media.py

```diff
@@ -0,0 +1,179 @@
+        media_content_type="root",
+        media_class=MediaClass.DIRECTORY,
+        children_media_class=MediaClass.DIRECTORY,
+        title="Jellyfin",
+        can_play=False,
+        can_expand=True,
+        children=children,
+    )
+
+
+async def build_item_response(
+    hass: HomeAssistant,
+    client: JellyfinClient,
+    user_id: str,
+    media_content_type: str | None,
+    media_content_id: str,
+) -> BrowseMedia:
+    """Create response payload for the provided media query."""
+    title, media, thumbnail = await get_media_info(
+        hass, client, user_id, media_content_type, media_content_id
+    )
+
+    if title is None or media is None:
+        raise BrowseError(f"Media not found: {media_content_type} / {media_content_id}")
+
+    children = await asyncio.gather(
+        *(item_payload(hass, client, user_id, media_item) for media_item in media)
+    )
+
+    response = BrowseMedia(
+        media_class=CONTAINER_TYPES_SPECIFIC_MEDIA_CLASS.get(
+            str(media_content_type), MediaClass.DIRECTORY
+        ),
+        media_content_id=media_content_id,
+        media_content_type=str(media_content_type),
+        title=title,
+        can_play=bool(media_content_type in PLAYABLE_MEDIA_TYPES and media_content_id),
+        can_expand=True,
+        children=children,
+        thumbnail=thumbnail,
+    )
+
+    response.calculate_children_class()
+
+    return response
+
+
+def fetch_item(client: JellyfinClient, item_id: str) -> dict[str, Any] | None:
+    """Fetch item from Jellyfin server."""
+    result = client.jellyfin.get_item(item_id)
+
+    if not result:
+        return None
+
+    item: dict[str, Any] = result
+    return item
+
+
+def fetch_items(
+    client: JellyfinClient,
+    params: dict[str, Any],
+) -> list[dict[str, Any]] | None:
+    """Fetch items from Jellyfin server."""
+    result = client.jellyfin.user_items(params=params)
+
+    if not result or "Items" not in result or len(result["Items"]) < 1:
+        return None
+
+    items: list[dict[str, Any]] = result["Items"]
+
+    return [
+        item
+        for item in items
+        if not item.get("IsFolder")
+        or (item.get("IsFolder") and item.get("ChildCount", 1) > 0)
+    ]
+
+
+async def get_media_info(
+    hass: HomeAssistant,
+    client: JellyfinClient,
+    user_id: str,
+    media_content_type: str | None,
+    media_content_id: str,
+) -> tuple[str | None, list[dict[str, Any]] | None, str | None]:
+    """Fetch media info."""
+    thumbnail: str | None = None
+    title: str | None = None
+    media: list[dict[str, Any]] | None = None
+
+    item = await hass.async_add_executor_job(fetch_item, client, media_content_id)
+
+    if item is None:
+        return None, None, None
+
+    title = item["Name"]
+    thumbnail = get_artwork_url(client, item)
+
+    if item.get("IsFolder"):
+        media = await hass.async_add_executor_job(
+            fetch_items, client, {"parentId": media_content_id, "fields": "childCount"}
```

Comment:
Is there a reason this is done as second step?

I'd assume the API rerturns the object with a missing field if it can't statisfy the childCount, and avoiding an API call should be preferrable to avoiding some bytes in other calls.

Reply:
the first call fetches the item by id, the second one gets children of item ID if type is a folder. I think before it was always making two calls based on if it was certain library type, I just use IsFolder instead.

let me check again if ChildCount is available when you query item and we could maybe save the second api call

Reply:
it seems the child count can be unreliable when you first call the get_item based on my install atleast. Not sure why but i have a few other bugs to report to jellyfin as well. We can always refine this behavior